### PR TITLE
ui: multi-select and batched gallery image operations

### LINKED ISSUE #3882: [enhancement]: multi-select in gallery would drastically improve usage.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Contact Details

_No response_

### What should this feature add?

I feel like the gallery really needs the ability to shift or ctl select multiple items do delete, especially after finishing generating a crapton of images i want to clean things up and delete a bunch, deleting one at a time seems frustrating

Would be even better if we could just box select them with mouse to multi-select

### Alternatives

_No response_

### Aditional Content

_No response_

### LINKED ISSUE #4088: [bug]: FOREIGN KEY constraint failed
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS

Windows

### GPU

cuda

### VRAM

16

### What version did you experience this issue on?

3.0.1

### What happened?

I can't upload a file into the image to image or the unified canvas. I get this error "[2023-07-30 21:43:45,244]::[InvokeAI]::ERROR --> Problem saving image record and file: FOREIGN KEY constraint failed"

### Screenshots

image

### Additional context

_No response_

### Contact Details

drriotmd@gmail.com

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No 😬 

## Description

[feat: add multi-select to gallery](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)
[c3e99ad](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)

multi-select actions include:
- drag to board to move all to that board
- right click to add all to board or delete all

backend changes:
- add routes for changing board for list of image names, deleting list of images
- change image-specific routes to `images/i/{image_name}` to not clobber other routes (like `images/upload`, `images/delete`)
- subclass pydantic `BaseModel` as `BaseModelExcludeNull`, which excludes null values when calling `dict()` on the model. this fixes inconsistent types related to JSON parsing null values into `null` instead of `undefined`
- remove `board_id` from `remove_image_from_board`

frontend changes:
- multi-selection stuff uses `ImageDTO[]` as payloads, for dnd and other mutations. this gives us access to image `board_id`s when hitting routes, and enables efficient cache updates.
- consolidate change board and delete image modals to handle single and multiples
- board totals are now re-fetched on mutation and not kept in sync manually - was way too tedious to do this
- fixed warning about nested `<p>` elements
- closes #4088 , need to handle case when `autoAddBoardId` is `"none"`
- add option to show gallery image delete button on every gallery image

frontend refactors/organisation:
- make typegen script js instead of ts
- enable `noUncheckedIndexedAccess` to help avoid bugs when indexing into arrays, many small changes needed to satisfy TS after this
- move all image-related endpoints into `endpoints/images.ts`, its a big file now, but this fixes a number of circular dependency issues that were otherwise felt impossible to resolve

## Related Tickets & Documents

- Supersedes #3669 #3771
- Closes #3882 #4088

## QA Instructions, Screenshots, Recordings

Basic functionality to review:
- Use shift and ctrl/cmd to select multiple images
- Drag and drop images between boards
- Right click multi-selection to change the board for a group of images
- Right click multi-selection to delete many images at once
- General drive-through of gallery features & auto-add

Specific cases:
- Deleting a single image should still work as before
- Deleting a single image that is in use in another feature should prompt you and reset that feature when deletion is confirmed
- Same as above for multiple images
- Adding a single image to a board should still work as before
- Adding multiple images to a board should work via context menu and dnd

There's a lot to test here and some complicated cache updates, so it's hard to test all the possible sequences of events.

## Added/updated tests?

- [ ] Yes
- [x] No 😬

### CODE REVIEW

File: invokeai/app/api/routers/board_images.py
Comment (psychedelicious) [inline invokeai/app/api/routers/board_images.py:1]: @maryhipp These are the changes needed for batch `board_image` operations. May be some additional changes but just giving you a heads up. The destination `board_id` is provided in each request - do we also need to provide the image's source/current `board_id`?

File: invokeai/frontend/web/src/services/api/endpoints/boards.ts [hunk 1]
```diff
@@ -82,6 +83,44 @@ export const boardsApi = api.injectEndpoints({
...
+        method: 'GET',
+      }),
+      providesTags: (result, error, arg) => [
+        { type: 'BoardAssetsTotal', id: arg ?? 'none' },
+      ],
+      transformResponse: (response: OffsetPaginatedResults_ImageDTO_) => {
+        return response.total;
+      },
+    }),
+
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/services/api/endpoints/boards.ts:123]: I tried manually keeping the board totals in sync but boy was it a pain. I'm just using the list images routes, with a `limit` of `0`, as a easy way to query for a board's image counts. 

We need separate queries for the `images` and `assets`, so while we could make dedicated routes for the counts, we'd still have to do a join. So I figure using the existing route is fine for now.

File: invokeai/frontend/web/tsconfig.json [hunk 2]
```diff
@@ -13,6 +13,8 @@
     "moduleResolution": "Node",
     // TODO: Disabled for IDE performance issues with our translation JSON
     // "resolveJsonModule": true,
+    "noUncheckedIndexedAccess": true,
+    "strictNullChecks": true,
```
Comment (psychedelicious) [inline invokeai/frontend/web/tsconfig.json:17]: Tightened up the reins a bit. Ran into some bugs related to indexed array access: TS does not type indexed array access values as a union with `undefined`, so it's very easy to end up with undefined when working with arrays.

Enabling this rule unions *all* indexed array access values with `undefined`. Many more runtime checks are needed, but no more errors.

File: invokeai/frontend/web/src/app/components/ImageDnd/typesafeDnd.tsx [hunk 3]
```diff
@@ -78,14 +82,12 @@ export type ImageDraggableData = BaseDragData & {
   payload: { imageDTO: ImageDTO };
 };
 
-export type ImageNamesDraggableData = BaseDragData & {
-  payloadType: 'IMAGE_NAMES';
-  payload: { image_names: string[] };
+export type ImageDTOsDraggableData = BaseDragData & {
+  payloadType: 'IMAGE_DTOS';
+  payload: { imageDTOs: ImageDTO[] };
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/app/components/ImageDnd/typesafeDnd.tsx:87]: I changed the drag and drop payload for multiple images to be an array of DTOs instead of just image names. This gives us more flexibility when handling these dnd events. 

For example, we now have access to an image's current `board_id` when dropping multiple images on a board. This allows us to greatly reduce the number of network requests needed when doing batched board management operations.

File: invokeai/frontend/web/scripts/typegen.js
Comment (psychedelicious) [inline invokeai/frontend/web/scripts/typegen.js:1]: There's no need for this to be a TS script. Changed to JS, runs a lot faster now (not like it was slow before tho).

File: invokeai/app/util/model_exclude_null.py
Comment (psychedelicious) [inline invokeai/app/util/model_exclude_null.py:1]: Weird issue with the OpenAPI types - the `ImageDTO`s were coming back with `null` as values for empty properties. For example, if an image was not on a board, it would have `board_id: null` instead of `board_id: undefined` or not having a `board_id` at all.

The generated types for eg `ImageDTO` do *not* have `null` in them anywhere - so we were expecting `board_id: string | undefined` but the actual object were `board_id: string | null`. This caused *really* confusing problems.

The solution is to override the pydantic `BaseModel`'s `dict()` method to use the `exclude_none` argument. Then the `ImageDTO`s simply omit empty props, and the objects we receive have the expected shape.

Instead of overriding `dict()` on every model that needs it, we can subclass `BaseModel` and have eg `ImageDTO` inherit from it.

File: invokeai/app/api/routers/board_images.py [hunk 4]
```diff
@@ -53,3 +67,66 @@ async def remove_board_image(
...
+    operation_id="create_board_images_from_list",
+    responses={
+        201: {"description": "Board images were created successfully"},
+    },
+    status_code=201,
+    response_model=CreateBoardImagesFromListResult,
+)
+async def create_board_images_from_list(
+    board_images: list[BoardImage] = Body(
+        description="The board images to create", embed=True
```
Comment (maryhipp) [inline invokeai/app/api/routers/board_images.py:82]: What is the use case for having a list of BoardImages vs a list of Images with one board_id? Are there front-end actions that can trigger this for multiple board/image combinations?

**Comment by maryhipp** [comment]
I have only skimmed this so far, but wondering what happens when a group of images are moved from board A to board B? I only see functionality to add or remove from board, not change boards.

File: invokeai/app/api/routers/board_images.py
Comment (psychedelicious) [reply invokeai/app/api/routers/board_images.py:1]: Ok, would it be helpful to provide a original board ids with their image names though? 

Good that it's not necessary but we now have access to them when request is made, so it's no trouble to include them.

File: invokeai/app/api/routers/board_images.py (hunk 4 above)
Comment (psychedelicious) [reply invokeai/app/api/routers/board_images.py:82]: Yes, a batch may have images from any board, and this allows it to support multi select in an intuitive way. Imagine you want to make a board out of your current batch

**Comment by psychedelicious** [comment]
@maryhipp Because images can only be on one board, creating a board image is equivalent to both moving an image from one board to another, and adding image to a board. 

We could add move as a distinct operation but there's substantially more logic to add, and it makes multiple image operations a lot more complex.

**Comment by maryhipp** [comment]
> @maryhipp Because images can only be on one board, creating a board image is equivalent to both moving an image from one board to another, and adding image to a board.
> 
> We could add move as a distinct operation but there's substantially more logic to add, and it makes multiple image operations a lot more complex.

OK - I am a bit wary of relying solely on a database constraint to dictate what is essentially business logic (that is my understanding of the implementation); I feel like it could lead to unexpected or unintended behavior if someone isn't aware of it or doesn't understand SQL. This is the only line that is enforcing this, right? `ON CONFLICT (image_name) DO UPDATE SET board_id = ?;`

File: invokeai/app/api/routers/board_images.py (hunk 4 above)
Comment (psychedelicious) [reply invokeai/app/api/routers/board_images.py]: I decided to just make this a list of image names with a single `board_id`, we can handle the other behavior if the actual need arises

File: invokeai/app/api/routers/board_images.py
Comment (psychedelicious) [reply invokeai/app/api/routers/board_images.py:1]: Sorry to back and forth on this. In the end, I've removed `board_id` from the `remove from board` routes and services.

**Comment by psychedelicious** [comment]
> OK - I am a bit wary of relying solely on a database constraint to dictate what is essentially business logic (that is my understanding of the implementation); I feel like it could lead to unexpected or unintended behavior if someone isn't aware of it or doesn't understand SQL. This is the only line that is enforcing this, right? `ON CONFLICT (image_name) DO UPDATE SET board_id = ?;`

Yep that's what handles changing the board if the image is already on another board. A more explicit handling might be to first remove images from their board before adding them, in the `add_images_to_board` route. Should behave exactly the same

**Comment by psychedelicious** [comment]
Ok, ready for review. Squashed:

[feat: add multi-select to gallery](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)

multi-select actions include:
- drag to board to move all to that board
- right click to add all to board or delete all

backend changes:
- add routes for changing board for list of image names, deleting list of images
- change image-specific routes to `images/i/{image_name}` to not clobber other routes (like `images/upload`, `images/delete`)
- subclass pydantic `BaseModel` as `BaseModelExcludeNull`, which excludes null values when calling `dict()` on the model. this fixes inconsistent types related to JSON parsing null values into `null` instead of `undefined`
- remove `board_id` from `remove_image_from_board`

frontend changes:
- multi-selection stuff uses `ImageDTO[]` as payloads, for dnd and other mutations. this gives us access to image `board_id`s when hitting routes, and enables efficient cache updates.
- consolidate change board and delete image modals to handle single and multiples
- board totals are now re-fetched on mutation and not kept in sync manually - was way too tedious to do this
- fixed warning about nested `<p>` elements
- closes https://github.com/invoke-ai/InvokeAI/issues/4088 , need to handle case when `autoAddBoardId` is `"none"`
- add option to show gallery image delete button on every gallery image

frontend refactors/organisation:
- make typegen script js instead of ts
- enable `noUncheckedIndexedAccess` to help avoid bugs when indexing into arrays, many small changes needed to satisfy TS after this
- move all image-related endpoints into `endpoints/images.ts`, its a big file now, but this fixes a number of circular dependency issues that were otherwise felt impossible to resolve

**Comment by psychedelicious** [comment]
I've updated the PR description with changes and how to test.

File: invokeai/frontend/web/src/services/api/endpoints/boards.ts (hunk 1 above)
Comment (blessedcoolant) [reply invokeai/frontend/web/src/services/api/endpoints/boards.ts:110]: This man reviewing his own PR. We can call it a day. :P

Comment (psychedelicious) [reply invokeai/frontend/web/src/services/api/endpoints/boards.ts:110]: Lol

File: invokeai/app/api/routers/images.py [hunk 5]
```diff
@@ -77,7 +78,7 @@ async def upload_image(
         raise HTTPException(status_code=500, detail="Failed to create image")
 
 
-@images_router.delete("/{image_name}", operation_id="delete_image")
+@images_router.delete("/i/{image_name}", operation_id="delete_image")
```
Comment (maryhipp) [inline invokeai/app/api/routers/images.py:81]: Why these routes changes?

File: invokeai/frontend/web/src/features/controlNet/components/parameters/ParamControlNetWeight.tsx [hunk 6]
```diff
@@ -1,32 +1,18 @@
...
+import { memo, useCallback } from 'react';
 
 type ParamControlNetWeightProps = {
-  controlNetId: string;
+  controlNet: ControlNetConfig;
 };
 
 const ParamControlNetWeight = (props: ParamControlNetWeightProps) => {
-  const { controlNetId } = props;
+  const { weight, isEnabled, controlNetId } = props.controlNet;
```
Comment (maryhipp) [inline invokeai/frontend/web/src/features/controlNet/components/parameters/ParamControlNetWeight.tsx:14]: All of these controlnet config changes are unrelated to multiselect, right?

**Comment by maryhipp** [comment]
I just tested locally and it looks great! Everything seems to be working really well. The only thing I noticed is that when you choose multiple images to delete using multiselect, there is a lengthy pause (this was with like 20 images) in the UI before anything happens. Maybe a small loading state or something would help, but just a nice-to-have.

**Review by maryhipp — APPROVED** [review]
Had some small questions but overall this looks 🔥 and I think we can always iterate on the small stuff

File: invokeai/app/api/routers/images.py (hunk 5 above)
Comment (psychedelicious) [reply invokeai/app/api/routers/images.py:81]: Concerned that an image name may clobber `images/upload` or `images/batch` (or any other routes we may add in the future). It really shouldn't happen bc the image names are UUIDv4, but this makes it impossible

File: invokeai/frontend/web/src/features/controlNet/components/parameters/ParamControlNetWeight.tsx (hunk 6 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/controlNet/components/parameters/ParamControlNetWeight.tsx:14]: Yes - I kept on running into bugs related to indexed array access that gave me `undefined` when I expected an `ImageDTO`, due to the array length. Typescript does not mark indexed access as possibly undefined by default.

I enabled `noUncheckedIndexedAccess` in `tsconfig` to make Typescript be stricter with this, which showed me all the places where this class of but could pop up. 

Unfortunately, it was a lot of places unrelated to the gallery, and I fixed all occurrences.

**Comment by psychedelicious** [comment]
> I just tested locally and it looks great! Everything seems to be working really well. The only thing I noticed is that when you choose multiple images to delete using multiselect, there is a lengthy pause (this was with like 20 images) in the UI before anything happens. Maybe a small loading state or something would help, but just a nice-to-have.

Assuming you mean after you confirm a multiple Delete - I couldn't figure out how to do optimistic updates for these actions without totally invalidating the caches, so we are waiting for the server response before updating the UI. That's the delay

This is the only un-ticked checkbox on the roadmap for this feature... we'd need to "freeze" all images pending mutation (hard) or freeze all image interactions (ugly) while we wait for the network request to finish. If a pause is tenable for now, I'd like to address this in a followup.

**Comment by maryhipp** [comment]
We could also just add a loader or something over the gallery during that delay - but agree it can be a follow-up

File: invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts [hunk 7]
```diff
@@ -1,57 +1,72 @@
...
+export const addRequestedSingleImageDeletionListener = () => {
   startAppListening({
     actionCreator: imageDeletionConfirmed,
     effect: async (action, { dispatch, getState, condition }) => {
-      const { imageDTO, imageUsage } = action.payload;
+      const { imageDTOs, imagesUsage } = action.payload;
 
-      dispatch(isModalOpenChanged(false));
+      if (imageDTOs.length !== 1 || imagesUsage.length !== 1) {
+        // handle multiples in separate listener
```
Comment (brandonrising) [inline invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts:23]: Any reason single and multi image deletion couldn't be the same listener and backend endpoint? Like, single image could just send a list of images of size 1. Just seems like this could be annoying to maintain(or forgotten about) in the future.

Comment (psychedelicious) [reply invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts:23]: I had approached the same conclusion but am a bit burnt out on the gallery stuff and would prefer to leave it for a future cleanup

File: invokeai/frontend/web/src/features/canvas/hooks/useColorUnderCursor.ts [hunk 8]
```diff
@@ -33,6 +33,10 @@ const useColorPicker = () => {
           1
         ).data;
 
+      if (!(a && r && g && b)) {
```
Comment (brandonrising) [inline invokeai/frontend/web/src/features/canvas/hooks/useColorUnderCursor.ts:36]: Even when it makes sense for variables to be 1 letter long, it still so PTSD inducing from my university math days lol

**Review by brandonrising — APPROVED** [review]
Overall lgtm, I tried pretty much everything in the UI that's affected by these changes and it all looked great. I left a few comments but I don't think any of them are blocking, feel free to ignore them or address them in a future PR. Great work!

The only other high level thing I thought of while reviewing this is that multi-select download could be really nice to have as well.

File: invokeai/app/api/routers/board_images.py [hunk 9]
```diff
@@ -29,26 +35,78 @@ async def create_board_image(
...
+    try:
+        added_image_names: list[str] = []
+        for image_name in image_names:
+            try:
+                ApiDependencies.invoker.services.board_images.add_image_to_board(
+                    board_id=board_id, image_name=image_name
+                )
+                added_image_names.append(image_name)
+            except:
+                pass
```
[the remark this answers is not in the document]
Comment (psychedelicious) [reply invokeai/app/api/routers/board_images.py:83]: Yeah, probably so. Raised in #4141

File: invokeai/frontend/web/src/features/canvas/hooks/useColorUnderCursor.ts (hunk 8 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/canvas/hooks/useColorUnderCursor.ts:36]: I promise there aren't many of these! This is only in the PR due to typescript refactoring, else you'd never have known

File: invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts (hunk 7 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts:23]: Raised in  #4142

**Comment by psychedelicious** [comment]
I've resolved merge conflicts related to #4086 and tested that the auto-assign on click still works. Thanks for the thorough review @brandonrising !
